### ISSUE TITLE
Support renaming existing movie folders.

### ISSUE THREAD

**Issue by chrisallen** [Radarr/Radarr#353]
Currently, when adding a new movie, Radarr will create the movie folder automatically and all is well.

In the case where a user has a folder for each movie, but the folder does not match the movies title, we could allow renaming it, which would keep all the files intact that reside in the file.

One use case would be if we wanted to allow moving from `The Name (year)\The Name (year).mkv` to  `Name, The (year)\The Name (year).mkv`

**Comment by galli-leo** [comment]
Did you try using the organize option?

**Comment by fedoranimus** [comment]
Reorganize should do this, be sure to run "Update Library" before doing so.

**Comment by Flippage** [comment]
Can confirm this doesn't work. Also doesn't affect srt or other subtitle files inside the folder.

Folder remains untouched by "Organise" in Movie Editor or "Preview Rename" on the movie itself.

Tried on a few different movies.

**Comment by fedoranimus** [comment]
We don't support handling subtitles right now - that's a separate issue.

I will look into folder renaming.

**Comment by Flippage** [comment]
Any update on this one @fedoranimus ?

I'm just waiting on the renaming of existing folders before I import the rest of my collection :)

**Comment by fedoranimus** [comment]
@Flippage The feature works, but has not been merged yet. I want to add some UI options to disable it as well as write more test coverage around it. It's a destructive process and can screw stuff up if not used wisely.

**Comment by crockadile35** [comment]
ok i'm in the right message here. i'm looking for a way to use the organize feature to rename like lets say about 500 movies instead of 1 by 1. is that possible?

**Comment by fedoranimus** [comment]
@crockadile35 Organize will rename the movie files, but not the folders containing them.

**Comment by crockadile35** [comment]
i'm ok with the that setup, when i click on bulk import movies and then select the storage where they are, i can only import 15 out a time. it can import 100, nothing happens

**Comment by fedoranimus** [comment]
Bulk Import is not a released feature and has several bugs which I'm currently working through.

**Comment by DejaVu** [comment]
Not sure why Movies Folder renaming would be a destructive process. 
Surely it's just renaming the recognised file name and using the Meta Data scrapped to create the parent folder name? 

**D:\Movies** is the Root of the Movie Folder (How Sonarr works for TV).
The current filename in **D:\Movies\New Folder\Fatal Attraction (1987).avi**

The required result _(as requested in the settings)_ under Media Management and/or Movie Renaming (I thought it was already doing it after looking in there thinking its implemented! 

D:\Movies\Fatal Attraction (1987)\Fatal Attraction (1987).avi

**Comment by dafzthomas** [comment]
Where can I find this "Organise" feature? @fedoranimus

**Comment by galli-leo** [comment]
@dafzthomas Movie Editor -> Organize

**Comment by SP3NGL3R** [comment]
This is the only thing currently in Radarr that's frustrating me :( ... +1 for sure.

**Comment by galli-leo** [comment]
This is actually in the latest release.

**Comment by SP3NGL3R** [comment]
Weird. I guess my build didn't update until yesterday. This appears to be here now. Previously when I hit the 'preview rename' icon it would pop with a blank list saying "all's good". But I do see it now. ... THANKS!!!

**Comment by greenxeyezz** [comment]
Tried updating library, then organize, radarr will NOT rename existing folders. Am I missing something?

**Comment by glardian** [comment]
@greenxeyezz
The UI doesn't tell you this, but for the folder rename to work you must set the new static path property to no. Unfortunately, I've been unable to get this property to change in bulk using the movie editor. I've had to do it individually per movie.

Also, the way this renamer has been implemented doesn't work as I'd have expected. It doesn't actually rename the folder, it creates a new folder and movies the movie file into it. This is hardly desirable. This creates a bunch of duplicate folders with any files other than the movie file (subtitles, artwork, etc.) orphaned in the original un-renamed folder. Maybe we can have a toggle if we understand the risks and want the folder renamed in place?

This request should really be reopened.

**Comment by MONKiCODE** [comment]
Any ETA on this feature getting implemented? I just need it to do what CouchPotato has been doing for organizing existing movie folders.

**Comment by galli-leo** [comment]
@MONKiCODE It is already implemented. For now subtitles and other extra files are not moved though, since we haven't implemented support for them yet.

**Comment by glardian** [comment]
@galli-leo Is there a reason this feature was designed to create a new folder and move the movie file into it? Would it not be better to rename the existing folder? As it stands, I'd have to go through and delete the old folders for ~1000 movies after the "rename". Renaming the existing folder in place would also prevent the subtitles, etc. from getting orphaned.

**Comment by nathan951** [comment]
> @galli-leo Is there a reason this feature was designed to create a new folder and move the movie file into it? Would it not be better to rename the existing folder? As it stands, I'd have to go through and delete the old folders for ~1000 movies after the "rename". Renaming the existing folder in place would also prevent the subtitles, etc. from getting orphaned.

Does it Rename the directory? or will it create new leaving orphaned scraped data in old folder?

**Comment by glardian** [comment]
> Does it Rename the directory (so all files in dir stay in tact)?
Or will it create new folder, leaving orphaned scraped data in old folder?

@nathan951 

It does NOT rename the directory. It creates a new folder and only moves the movie file to the new folder. It leaves an orphaned folder with any other files that you need to manual move or delete. This has been implemented poorly.

**Comment by jonathan987** [comment]
I use, Batch Rename Files Tool. You can easily found hier go to google.com and type BatchRenameFiles and check the first result that allows you to quickly rename all the files in a specified directory.

**Comment by kormic911** [comment]
+1 this feature should really rename the folder and not make a new one and copy the file. This seems to be a major oversight on the implementation of this feature which causes unneeded headaches.

**Comment by sho-87** [comment]
curious about this too. is actual renaming (rather than copying and orphaning) planned for v1?

**Comment by galli-leo** [comment]
@sho-87 Idk, depends on how much demand there is for this feature. Because for now, a lot of people seem content with the "buggy" solution we currently have. Also it probably won't be actual renaming for the final solution but rather fully dynamic paths. i.e. Radarr will either have a predefined static path (like it is now, with disk scan service, etc.) or it will just throw the media file at the path it just calculated (no disk scan service, etc.) and forget about it.

**Comment by sho-87** [comment]
@galli-leo I think actual renaming/moving of additional files benefits various types of users:

1) Those that have an existing media library. This feature would make the transition to radarr easier, whether youre coming from couchpotato or manually managing movies. all of my movies were managed by couchpotato before moving to radarr and the files and folders were named like `<movie>-<data>-<quality>`. If, for whatever reason (e.g. change the naming scheme of my movies), I wanted to change the folder names and have radarr manage the movies I would end up with two folders in addition to orphaned files. Of course I could manually rename the folder and then point radarr to the new directory, but thats not feasible if you have a lot of movies. Furthermore, radarr (currently) only likes things in the `<movie>-<year>` format so if you have media with any other naming scheme you're in a bit of trouble. Also, with an existing library the folders already contain nfo/subs/images and its a bit redundant to grab all of that stuff again instead of radarr just moving everything across. 

2) Those that want to use radarr for movie management and dont have a media server. I run Emby as a media server and Im not sure if it can move subtitles - it downloads them, but if I already have sub files then it seems more sensible to just move them. This is made worse if you use a subtitle service that has a daily limit on downloads (e.g. open subtitles). Further, I'm sure not everyone has a media server running Emby/Plex, and those people just want something to manage the movie files rather than to serve them. To me, management includes the moving of movies and associated files

3) Users that are new to the world of media servers and media management. I think its important to not break the user's expectation of how things work, especially if theyre new to this type of thing. Violating expectations is the cause of much frustration and many github issues. Especially in this case where (at least to me) it seems like the most natural thing for folders to be renamed rather than to create and move. It feels a bit uncomfortable to see an option like "Paths default to static" and not wanting to turn it off because its unclear whether it will do what I expect, or mess up my folder structure/naming in some way

Also, Im not sure if fully dynamic paths would address the core issue. If I tell radarr to rename a folder using dynamic paths, would I still end up with orphaned files, or multiple folders?

So, I'm still in favor of some folder renaming solution. It feels like a big quality of life improvement and would reduce frustration/misunderstanding/having to turn to external tools to rename.

Anyway, thats just my 2 cents. I realize radarr is still technically in development and I'm loving using it so far. I guess I just dont have the technical knowledge to know _why_ it currently works this way

**Comment by galli-leo** [comment]
@sho-87 Regarding 3., new users don't need to use dynamic paths anyways (60% of Radarr users don't even know what it is). Though I mostly agree with the rest of your points.

Regarding the why. We can only "rename" the folder, once the movie file is imported (into the new dir), in which case it's already "too late".

Anyways, I found out why it wasn't removing empty folders (as it should) and also just added a few lines to make it copy all other files in the dir, then delete it. I think this should solve it for almost all users.

**Comment by DesktopMan** [comment]
@galli-leo Question on how this currently works with folder renaming: Isn't it odd to rename folders on "Update Library"? Wouldn't it be better to rename folders when you organize a movie (or multiple movies)?

**Comment by JesseWV** [comment]
This is a feature I've been waiting to be properly implemented for awhile. In my opinion it should be an "Organize" button on the Movie Editor page. Obviously it should rename the folder and ultimately rename any subtitle or other files in the folder when subtitle support is added.

**Comment by galli-leo** [comment]
@DesktopMan @JesseWV I agree it's odd to rename folders on update library, but this is mostly a temporary mesure until we correctly support "dynamic folders".

@JesseWV Organize already works for movie files, since subtitle support is added it should work for them as well IIRC.

**Comment by cani1989** [comment]
Any update on this?

**Comment by achmetinternet** [comment]
+1

**Comment by acosmichippo** [comment]
Hello,

I've been trying to rename my folders in the {Movie TitleThe} ({Release Year}), but have been unsuccessful.

I have "Automatically rename folders" on and "movie paths default to static" turned off.

From what I understand, the folders should be renamed when I click "update library" on the home page, but it doesn't happen.  The library scan runs, but no folders are renamed.  I've also tried doing it for individual movies with the organize and rename tool, but that doesn't do it either.

Is there some other setting I need to change for this to work?

here is an example in the logs from the library update:

>18-10-18 16:40:06.4|Info|RefreshMovieService|Updating Info for The Grand Budapest Hotel
>18-10-18 16:40:06.7|Info|DiskScanService|Scanning disk for The Grand Budapest Hotel
>18-10-18 16:40:06.7|Info|DiskScanService|Completed scanning disk for The Grand Budapest Hotel
>18-10-18 16:40:06.7|Info|ExistingMetadataImporter|Found 0 existing metadata files
>18-10-18 16:40:06.7|Info|ExistingSubtitleImporter|Found 0 existing subtitle files
>18-10-18 16:40:06.7|Info|ExistingOtherExtraImporter|Found 0 existing other extra files
>18-10-18 16:40:06.7|Info|ExistingExtraFileService|Found 0 extra files


Thank you for any help!

**Comment by JohnAustin16** [comment]
I had the same problem and with 800 some movies renaming folders manually was not an option. However, my need to be neat took over and I spent some time with python and came up with this attached script.
Essentially the script will go through a folder (recursively) of your choosing and find all files with: .mp4 .mkv .avi .wmv .h264 extension it then moves them into a folder that’s named identical to the movie file. I used Radarr to rename all my movies then ran this script to rename all the folders. This may not be for everyone since it will only move the movies and nothing else in the folder but it might be a building block if there’s any other python enthusiast on here. 
Have a look at the code and if you like it feel free to use it.
PS: I am not responsible for any mishaps due to this code use at your own risk! 

[Folder_Renamer.zip](https://github.com/Radarr/Radarr/files/3636355/Folder_Renamer.zip)
